{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "103f5032",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9f57c494",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "aac1bfc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir = Path(\"/zdata/user-data/noam/data/p2cs/\")\n",
    "full_genomes_dir = base_dir / \"full_genomes_new\"\n",
    "organism_merged_genomes_dir = base_dir / \"organism_merged_genomes_new\"\n",
    "genome_analysis_dir = base_dir / \"genome_analysis_results\"\n",
    "mash_analysis_dir = base_dir / \"mash_analysis\"\n",
    "data_dir = mash_analysis_dir / \"distance_matrices\"  # For genome-level matrices\n",
    "taxonomy_dir = base_dir / \"taxonomy\"\n",
    "k_values = [11, 15, 19, 21]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "536e1be4",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3ecbfdc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "p2cs_filtered_groups_path = base_dir / \"merged_p2cs_data\" / \"p2cs_filtered_groups.pkl\"\n",
    "\n",
    "p2cs_filtered_groups = pd.read_pickle(p2cs_filtered_groups_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1c629fbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Original description</th>\n",
       "      <th>class</th>\n",
       "      <th>type</th>\n",
       "      <th>P2CS description</th>\n",
       "      <th>tm</th>\n",
       "      <th>file_name</th>\n",
       "      <th>...</th>\n",
       "      <th>organism</th>\n",
       "      <th>Gene_num</th>\n",
       "      <th>frame</th>\n",
       "      <th>proximity_group</th>\n",
       "      <th>tcs_organization</th>\n",
       "      <th>tcs_organization_int</th>\n",
       "      <th>proximity_group_size</th>\n",
       "      <th>num_cognates</th>\n",
       "      <th>aa_sequence</th>\n",
       "      <th>nt_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Asuc_0781</td>\n",
       "      <td>841792</td>\n",
       "      <td>842496</td>\n",
       "      <td>+</td>\n",
       "      <td>two component transcriptional regulator</td>\n",
       "      <td>RR</td>\n",
       "      <td>OmpR</td>\n",
       "      <td>Response regulator, OmpR family contains 1 Res...</td>\n",
       "      <td>0</td>\n",
       "      <td>ActsuDB_Actinobacillus_succinogenes_130Z</td>\n",
       "      <td>...</td>\n",
       "      <td>Actinobacillus succinogenes 130Z</td>\n",
       "      <td>781</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>Pair</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MTKILLVDDDIELTDLLGELLSLEGFDVVTAQNGLEALEKLDDGIC...</td>\n",
       "      <td>ATGACGAAAATTTTATTAGTCGATGACGATATCGAGTTAACTGATC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene   Start    Stop Strand                     Original description  \\\n",
       "0  Asuc_0781  841792  842496      +  two component transcriptional regulator   \n",
       "\n",
       "  class  type                                   P2CS description  tm  \\\n",
       "0    RR  OmpR  Response regulator, OmpR family contains 1 Res...   0   \n",
       "\n",
       "                                  file_name  ...  \\\n",
       "0  ActsuDB_Actinobacillus_succinogenes_130Z  ...   \n",
       "\n",
       "                           organism Gene_num  frame  proximity_group  \\\n",
       "0  Actinobacillus succinogenes 130Z      781      1                2   \n",
       "\n",
       "   tcs_organization tcs_organization_int  proximity_group_size  num_cognates  \\\n",
       "0              Pair                  2.0                     2           1.0   \n",
       "\n",
       "                                         aa_sequence  \\\n",
       "0  MTKILLVDDDIELTDLLGELLSLEGFDVVTAQNGLEALEKLDDGIC...   \n",
       "\n",
       "                                         nt_sequence  \n",
       "0  ATGACGAAAATTTTATTAGTCGATGACGATATCGAGTTAACTGATC...  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p2cs_filtered_groups.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6210893",
   "metadata": {},
   "source": [
    "# Create & Save Fasta's"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4556c3fc",
   "metadata": {},
   "source": [
    "**Test Amino acid fasta file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d6840eb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fasta file for hk filtered groups saved to /zdata/user-data/noam/data/p2cs/merged_p2cs_data/fasta/mini_sample_hk_filtered_groups.faa\n",
      "Fasta file for rr filtered groups saved to /zdata/user-data/noam/data/p2cs/merged_p2cs_data/fasta/mini_sample_rr_filtered_groups.faa\n"
     ]
    }
   ],
   "source": [
    "header_cols = ['Gene', 'organism', 'type', 'proximity_group_size', 'num_cognates']\n",
    "data_col = 'aa_sequence'\n",
    "\n",
    "# Create fasta files for hk's and for rr's\n",
    "for class_type in ['hk', 'rr']:\n",
    "    class_df = p2cs_filtered_groups[p2cs_filtered_groups['class'] == class_type.upper()].sample(200).copy()\n",
    "\n",
    "    fasta_path = base_dir / \"merged_p2cs_data\" / \"fasta\" / f\"mini_sample_{class_type}_filtered_groups.faa\"\n",
    "    with open(fasta_path, 'w') as f:\n",
    "        for index, row in class_df.iterrows():\n",
    "            header_values = [str(row[col]) for col in header_cols]\n",
    "            header_str = \" \".join(header_values)\n",
    "            f.write(f\">{header_str}\\n{row[data_col]}\\n\")\n",
    "\n",
    "    print(f\"Fasta file for {class_type} filtered groups saved to {fasta_path}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f784625",
   "metadata": {},
   "source": [
    "**Amino acid fasta files**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8487576c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fasta file for hk filtered groups saved to /zdata/user-data/noam/data/p2cs/merged_p2cs_data/fasta/hk_filtered_groups.faa\n",
      "Fasta file for rr filtered groups saved to /zdata/user-data/noam/data/p2cs/merged_p2cs_data/fasta/rr_filtered_groups.faa\n"
     ]
    }
   ],
   "source": [
    "header_cols = ['Gene', 'organism', 'type', 'proximity_group_size', 'num_cognates']\n",
    "data_col = 'aa_sequence'\n",
    "\n",
    "# Create fasta files for hk's and for rr's\n",
    "for class_type in ['hk', 'rr']:\n",
    "    class_df = p2cs_filtered_groups[p2cs_filtered_groups['class'] == class_type.upper()]\n",
    "\n",
    "    fasta_path = base_dir / \"merged_p2cs_data\" / \"fasta\" / f\"{class_type}_filtered_groups.faa\"\n",
    "    with open(fasta_path, 'w') as f:\n",
    "        for index, row in class_df.iterrows():\n",
    "            header_values = [str(row[col]) for col in header_cols]\n",
    "            header_str = \" \".join(header_values)\n",
    "            f.write(f\">{header_str}\\n{row[data_col]}\\n\")\n",
    "\n",
    "    print(f\"Fasta file for {class_type} filtered groups saved to {fasta_path}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63a4d3a",
   "metadata": {},
   "source": [
    "**Nucleotide fasta files**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ca435bca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fasta file for hk filtered groups saved to /zdata/user-data/noam/data/p2cs/merged_p2cs_data/fasta/hk_filtered_groups.fna\n",
      "Fasta file for rr filtered groups saved to /zdata/user-data/noam/data/p2cs/merged_p2cs_data/fasta/rr_filtered_groups.fna\n"
     ]
    }
   ],
   "source": [
    "header_cols = ['Gene', 'organism','type', 'proximity_group_size', 'num_cognates']\n",
    "data_col = 'nt_sequence'\n",
    "\n",
    "# Create fasta files for hk's and for rr's\n",
    "for class_type in ['hk', 'rr']:\n",
    "    class_df = p2cs_filtered_groups[p2cs_filtered_groups['class'] == class_type.upper()]\n",
    "\n",
    "    fasta_path = base_dir / \"merged_p2cs_data\" / \"fasta\" / f\"{class_type}_filtered_groups.fna\"\n",
    "    with open(fasta_path, 'w') as f:\n",
    "        for index, row in class_df.iterrows():\n",
    "            header_values = [str(row[col]) for col in header_cols]\n",
    "            header_str = \" \".join(header_values)\n",
    "            f.write(f\">{header_str}\\n{row[data_col]}\\n\")\n",
    "\n",
    "    print(f\"Fasta file for {class_type} filtered groups saved to {fasta_path}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10 (Cursor)",
   "language": "python",
   "name": "python310-cursor"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
